docs/adr/2023-11-07_029-streaming-persistence.md

```diff
@@ -0,0 +1,40 @@
+---
+slug: 29
+title: |
+  29. EventServer abstraction for event persistence
+authors: [@cardenaso11, @quantumplation]
+tags: []
+---
+
+## Status
+Draft
+
+## Context
+
+The Hydra node represents a significant engineering asset, providing layer 1 monitoring, peer to peer consensus, durable persistence, and an isomorphic Cardano ledger. Because of this, it is being eyed as a key building block not just in Hydra based applications, but other protocols as well.
+
+One remaining difficulty in integrating Hydra into a fully productionalized software stack is the persistence model.  Currently, the Hydra node achieves durability by writing a sequence of "StateChanged" events to disk. If a node is interrupted, upon restarting, it can replay just these events, ignoring their corresponding effects, to recover the same internal state it had when it was interrupted. However, downstream consumers don't have the same luxury.
```

Comment:
Would be good to complete this paragraph and fully motivate the change with _one_ example of how this would be useful in production, if possible from your experience

Reply:
There are more aspects to your analysis of approaches. IMO approach 1 would be most favorable because it does not increase the scope to be maintained and allows for more loose coupling with a clear interface. For example, I do not see it favorable to add AWS dependencies to the hydra-node just for optional storing on S3 (or kafka deps, or ...). Not sure about a plugin mechanism either, but maybe an option.

---

I do understand your frustration. Maybe we should use less ambiguous modes of communication like drawing/amending an architecture diagram to clarify one or the other approach?

Or prototyping what you actually need for your solution and chat about code?

Anyhow, please see my comments as early as possible feedback and we can also merge the ADR as is.. the actual implementation will result in discussions about details and further refinement anyhow.

Reply:
> For example, I do not see it favorable to add AWS dependencies to the hydra-node just for optional storing on S3 (or kafka deps, or ...)

Agreed, we do not plan to add this directly to the hydra node; the goal is to make it possible to add in a fork (or through a plugin mechanism), while at the same time reducing the overall burden of keeping that fork up to date.

As it is now, we'd have to make enough changes to the hydra code base that each time you touched the persistence mechanism at all, we'd have to go through a potentially painful merge process.

By having this mechanism, us (and other users) would simply be implementing a standard interface in the fork, and the chance of merge conflicts is much less drastic.